endpoint/OBDComp/emulator/obd_emulator.c

```diff
@@ -0,0 +1,171 @@
+/*
+ * Copyright (C) 2019-2020 BiiLabs Co., Ltd. and Contributors
+ * All Rights Reserved.
+ * This is free software; you can redistribute it and/or modify it under the
+ * terms of the MIT license. A copy of the license can be found in the file
+ * "LICENSE" at the root of this distribution.
+ */
+
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <time.h>
+#include <unistd.h>
+
+#include "common/logger.h"
+#include "endpoint/OBDComp/can-bus/can-utils.h"
+#include "obd_emulator.h"
+
+static int service_01_response(int fd, uint8_t pid);
+static int service_09_response(int fd, uint8_t pid);
+
+static void print_help(void) { printf("Usage: obd_emulator <can_interface>\n"); }
+void emulator_exit(int status, void* arg) {
+  int can_fd = *(int*)arg;
+  can_close(can_fd);
+}
+
+int main(int argc, char* argv[]) {
+  char* default_can_device = OBD_EMULATOR_INTERFACE;
+  if (argc < 2 || strncmp(argv[1], "-h", strlen("-h")) == 0) {
+    print_help();
+    exit(EXIT_SUCCESS);
+  }
+
+  if (argc >= 2 && argv[1] != NULL) {
+    default_can_device = argv[1];
+  }
+
+  int can_fd;
+  status_t ret;
+  ret = can_open(default_can_device, &can_fd);
+  if (ret != SC_OK) {
+    fprintf(stderr, "%s\n", ta_error_to_string(ret));
+    exit(EXIT_FAILURE);
+  }
+
+  printf("OBDII emulator start!\n");
+  on_exit(emulator_exit, (void*)&can_fd);
+
+  struct can_frame frame;
+  do {
+    ret = can_recv(can_fd, &frame);
+    if (ret != SC_OK) {
+      fprintf(stderr, "%s\n", ta_error_to_string(ret));
+      continue;
+    }
+    printf("0x%03X [%d] ", frame.can_id, frame.can_dlc);
+    for (int i = 0; i < frame.can_dlc; i++) printf("%02X ", frame.data[i]);
+    printf("\r\n");
+
+    if (frame.can_id == OBD_BROADCAST_ID) {
+      printf("Received OBD query\r\n");
+      if (frame.can_dlc >= 2) {
+        switch (frame.data[1]) {
+          case 1:
+            printf("Service 01: Show current data\n");
+            service_01_response(can_fd, frame.data[2]);
+            break;
+          case 9:
+            printf("Service 09: Show current data\n");
+            service_09_response(can_fd, frame.data[2]);
+            break;
+          default:
+            printf("The service 0x%02X is not support\n", frame.data[1]);
+            break;
+        }
+      } else {
+        printf("Error, frame too short. DLC = %d bytes.\n", frame.can_dlc);
+      }
+    }
+  } while (1);
+  can_close(can_fd);
+  return 0;
+}
+
+static int rand_int(int max, int min) {
```

Comment:
Avoid using `max` and `min`, which might be expanded by predefined macros. Instead, use upper/lower bound.